File: 402-remove-k-digits/remove-k-digits.py
class Solution(object):
    def removeKdigits(self, num, k):
        x = []
        for i in num:
            while k > 0 and x and x[-1] > i:
                x.pop()
                k -= 1
            x.append(i)
        
        while k > 0:
            x.pop()
            k -= 1
        
        return ''.join(x).lstrip('0') or '0'

File: 402-remove-k-digits/remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        List<Character> x = new ArrayList<>();
        for (char c : num.toCharArray()) {
            while (k > 0 && !x.isEmpty() && x.get(x.size() - 1) > c) {
                x.remove(x.size() - 1);
                k--;
            }
            x.add(c);
        }
        
        while (k > 0) {
            x.remove(x.size() - 1);
            k--;
        }
        
        StringBuilder sb = new StringBuilder();
        for (char c : x) {
            sb.append(c);
        }
        
        String result = sb.toString().replaceFirst("^0+(?!$)", "");
        return result.isEmpty() ? "0" : result;
    }
}

